File: sorting/491-minEatingSpeed.cpp
#include<vector>
#include<iostream>
#include<algorithm>


using VEC=std::vector<int>;

bool check(const VEC& piles,int h,int r)
{
    int ans=0;
    for(auto &p:piles)
    {
        if((p%r)!=0)
            ans+=(p/r)+1;
        else
            ans+=(p/r);
    }
    if(ans<=h)
        return true;
    return false;
}

int minEatingSpeed(const VEC& piles,int h)
{
    int start=1;
    int max=*std::max_element(piles.begin(),piles.end());
    int end=max;
    while(start<end)
    {
        int mid=start+(end-start)/2;
        if(check(piles,h,mid))
            end=mid;
        else
            start=mid+1;
    }
    return end;  //here also can return start, as they are same values
}

int main()
{
    VEC piles{30,11,23,4,20};//{3,6,7,11};//{30,11,23,4,20};
    int h=5;//8;//6;
    std::cout<<minEatingSpeed(piles,h);
}

File: sorting/491-minEatingSpeed.py
from math import ceil
class Solution(object):
    def minEatingSpeed(self, piles, h):
        """
        :type piles: List[int]
        :type h: int
        :rtype: int
        """

        # ans=0
        def f(r):
            return sum(ceil(i/r) for i in piles)
        M=max(piles)
        L=1
        R=M
        while L<R:
            mid=(L+R)//2
            if f(mid)<=h:
                R=mid
            else:
                L=mid+1
        return L

piles=[30,11,23,4,20]

h=6
# 6
print(Solution().minEatingSpeed(piles,h))